### ISSUE TITLE
Implement Strikeout/Strikethrough.

### ISSUE THREAD

**Issue by ILOVEPIE** [SABRE-JS/SABRE.js#64]
Draw strikeout/strikethrough on text.

**Comment by unclebinary1001** [comment]
Can I work on this?

**Comment by ILOVEPIE** [comment]
Sure, i'd read through src/canvas-2d-text-renderer.js as that's where your going to be implementing it. I'd also look into how to find the text midline. if you need help let me know.

**Comment by ILOVEPIE** [comment]
Seeing as I haven't heard back from you, I'll assume you aren't working on this and delete these messages so other people see it as open to work on.

**Comment by unclebinary1001** [comment]
I was planning on getting started with it this afternoon, my time. Can you
give me a day to setup and then get back to you if its something I can do?

**Comment by unclebinary1001** [comment]
@ILOVEPIE Thank you for your patience with me. I spent time reading through `src/canvas-2d-text-renderer.js ` to understand the existing code base. 

I set up a small hmtl/js playground on my machine to try to find the rendered text middle height. 

I wrote this code snippet which works well with finding the y-value for the rendered text  middle height and then drawing a strike through the rendered text:

```
let txt = "Hello World";
ctx.font = "60px Arial";
let txt_x = 40;
let txt_y = 270;
ctx.fillText(txt, txt_x, txt_y);

let txt_width = ctx.measureText(txt).width;

let txt_height = ctx.measureText(txt).actualBoundingBoxAscent;
let mid_txt_height = txt_height /2;

ctx.lineWidth = 4;
ctx.moveTo(txt_x, txt_y -  mid_txt_height);
ctx.lineTo(txt_x + txt_width, txt_y - mid_txt_height);
ctx.stroke();
```

I believe I can work on this issue. My next steps will be to write a function  for this, and  add a bool property to enable/disable strike through to `src/canvas-2d-text-renderer.js `

**Comment by ILOVEPIE** [comment]
The strikethrough property is already there in the overrides and styles.
The rendering code isn’t implemented.

>

**Comment by unclebinary1001** [comment]
Oh okay. I see it now. I also see the comment for Strikeout. 
Thanks.

**Comment by unclebinary1001** [comment]
Then, based on the code snippet I shared, would you advise that I write a new function for `draw_Strikeout` or add it to one of the existing functions?

**Comment by ILOVEPIE** [comment]
Also, another note, the code you posted won't work due to the code using textBaseline = "top" so you'll need to use this code instead 

```JavaScript
let mid_txt_height = (((measurements.actualBoundingBoxDescent??measurements.height)+(measurements.actualBoundingBoxAscent??0))/2 - (measurements.actualBoundingBoxAscent??0));
```

and then change the subtractions to additions for adding the lmidpoint to the position.

**Comment by ILOVEPIE** [comment]
> Then, based on the code snippet I shared, would you advise that I write a new function for `draw_Strikeout` or add it to one of the existing functions?

I would create a new function similar to how ``_drawTextWithRelativeKerning`` works, the function should have a name like ``_drawStrikeThrough`` or ``_drawStrikeOut``.

**Comment by ILOVEPIE** [comment]
you should then call the function if strikethrough is enabled in the renderEvent function. Actually, now that I think about it, you'll need two functions, ``_drawStrikeThrough`` and ``_drawStrikeThroughWithRelativeKerning``.

**Comment by unclebinary1001** [comment]
So just an update, I was able to generate strike through that supports both stroke and fill for top textbaseline.

I am now working on incorporating relative kerning.

**Comment by unclebinary1001** [comment]
> 

So I tried this out, but the strike out was slightly below the midpoint of the letters. I tried to draw strike out line using `fillRect` and `strokeRect` instead of drawing a line.

**Comment by unclebinary1001** [comment]
> Also, another note, the code you posted won't work due to the code using textBaseline = "top" so you'll need to use this code instead
> 
> ```js
> let mid_txt_height = (((measurements.actualBoundingBoxDescent??measurements.height)+(measurements.actualBoundingBoxAscent??0))/2 - (measurements.actualBoundingBoxAscent??0));
> ```
> 
> and then change the subtractions to additions for adding the midpoint to the position.

So I tried this out, but the strike out was slightly below the midpoint of the letters. I tried to draw strike out line using fillRect and strokeRect instead of drawing a line.

**Comment by unclebinary1001** [comment]
Question. 

Do I add `_drawStrikeThrough` everytime `_drawTextWithRelativeKerning` occurs ofcourse with a check for bool `strikeOut`?

**Comment by ILOVEPIE** [comment]
you have to add it when there is a reference to drawing text without
kerning. when there is a reference to drawing text with kerning you need to
use the one for kerning.

**Comment by ILOVEPIE** [comment]
Also the reason why that code didn’t work in your test is because your test
and by code are different. They use different settings for the canvas so
therefore they’re gonna produce different output.

**Comment by unclebinary1001** [comment]
Oh, I see. Is there a way I can test the code in SABRE.js locally?

**Comment by ILOVEPIE** [comment]
make run

you’ll need a .ass file that uses strikeout which you can make using
Aegisub.

**Comment by unclebinary1001** [comment]
@ILOVEPIE I was able to generate a .ass file with Aegisub that uses strikeout: I selected the strikeout checkbox when editing the text then I exported the .ass file. 

I paste my .ass file content into `test/index.html` replacing the previous .ass in `'subs` content. I ran `make run` but I was still seeing the original subtitle content. 

I think there's something I am doing wrong. I am kindly asking for your help.
